framework/src/functions/PiecewiseConstantFromCSV.C

```diff
@@ -0,0 +1,115 @@
+  // element that is on the other side of a domain boundary
+  params.set<unsigned short>("ghost_layers") = 1;
+
+  params.addClassDescription("Uses data read from CSV to assign values");
+  return params;
+}
+
+PiecewiseConstantFromCSV::PiecewiseConstantFromCSV(const InputParameters & parameters)
+  : Function(parameters),
+    _read_prop_user_object(nullptr),
+    _column_number(getParam<unsigned int>("column_number")),
+    _read_type(getParam<MooseEnum>("read_type").getEnum<ReadTypeEnum>()),
+    _point_locator(_ti_feproblem.mesh().getPointLocator())
+{
+  if (_column_number < _ti_feproblem.mesh().dimension() && _read_type == ReadTypeEnum::VORONOI)
+    mooseWarning(
+        "The column requested in the function is likely to just be containing point coordinates");
+}
+
+void
+PiecewiseConstantFromCSV::initialSetup()
+{
+  // Get a pointer to the ElementPropertyReadFile. A pointer is used because the UserObject is not
+  // available during the construction of the function
+  _read_prop_user_object = &getUserObject<PropertyReadFile>("read_prop_user_object");
+
+  if (_read_type != _read_prop_user_object->getReadType())
+    paramError("read_type", "The PropertyReadFile UO should have the same read_type parameter.");
+}
+
+Real
+PiecewiseConstantFromCSV::value(Real, const Point & p) const
+{
+  if (_read_type == ReadTypeEnum::ELEMENT || _read_type == ReadTypeEnum::BLOCK)
+  {
+    // This is somewhat inefficient, but it allows us to retrieve the data in the
+    // CSV by element or by block.
+    std::set<const Elem *> candidate_elements;
+    (*_point_locator)(p, candidate_elements);
+
+    // Find the element with the lowest ID (that is also on the domain ???????)
+    unsigned int min_elem_id = std::numeric_limits<unsigned int>::max();
+    const Elem * min_id_elem;
+    unsigned int num_candidates = 0;
+
+    std::set<const Elem *>::iterator it = candidate_elements.begin();
+    while (it != candidate_elements.end())
+    {
+      if ((*it)->id() < min_elem_id)
+      {
+        min_elem_id = (*it)->id();
+        min_id_elem = (*it);
+        num_candidates++;
+      }
+      it++;
+    }
+
+    if (num_candidates == 0)
+      mooseError("No element located at ", p, " to search in element or block sorted CSV values");
+
+    return _read_prop_user_object->getData(min_id_elem, _column_number);
+  }
+  else if (_read_type == ReadTypeEnum::NODE)
+  {
+    // Get the node id
+    const auto node = _point_locator->locate_node(p);
+
+    if (!node)
+      mooseError("No node was found at", p, " for retrieving nodal data from CSV.");
+
+    return _read_prop_user_object->getNodeData(node, _column_number);
+  }
+  else if (_read_type == ReadTypeEnum::VORONOI)
+    // No need to search for the element if we're just looking at nearest neighbors
+    return _read_prop_user_object->getVoronoiData(p, _column_number);
+  else
+  {
+    mooseError("This should not be reachable. Implementation error somewhere");
+    return 0;
+  }
```

Comment:
```suggestion
  mooseError("This should not be reachable. Implementation error somewhere");
```

mooseError is `[[noreturn]]` so you just erroring is fine